src/PoolManager.sol

```diff
@@ -185,6 +193,8 @@ contract PoolManager is IPoolManager, Fees, NoDelegateCall, Claims {
         IPoolManager.ModifyPositionParams memory params,
         bytes calldata hookData
     ) external override noDelegateCall onlyByLocker returns (BalanceDelta delta) {
+        CurrentHookAddress.set(address(key.hooks));
```

Comment:
Hm I thought it wasn't necessary because you don't need a lock to call initialize, but maybe we should?

Reply:
Hm this just made me realize that there is actually strange behavior (not a security concern bc I think things will revert) but if a hook calls into any of these functions and there is NO current locker then the deltas accrue to the 0 address. Maybe we want to restrict that there is at least one locker? 

Secondly there are two scenarios Im thinking of 

1. If we don't set current hook in initialize and someone calls initialize with no lock -  The hook will have to acquire a lock to call any of the functions.
2. If we don't set current hook in initialize and someone has a lock already but calls initialize.  The hook can't operate on the outside lock.
3. If we do set current hook in initialize and someone calls initialize with no lock.  the hook can operate on the 0 address lock meaning any deltas applied will auto revert.
4. 3. If we do set current hook in initialize and someone calls initialize in their current lock. the hook can operate on the current lock.

Maybe 4 is the most desired? But we need to prevent 3 I think.. unless we're ok with that strange revert behavior

Reply:
Actually, this is a critical bug.

In the current code, a hook that has an access lock permission that has successfully set the LAST currentHook address to itself can call modifyPosition/swap/donate/mint/take successfully without settling deltas because it can call these functions OUTSIDE the lock. 

I've reproduced the vulnerability with this test here.

I think there are two things we can do
a) require there must be a lock ie in onlyByLocker we can check that if msg.sender is not the current locker and if the current locker == addr(0) we revert

b) at the end of a lock clear the currentHook address.

My gut is to actually do both to be safe but I think a) alone would solve the issue.

B) by itself I think actually does not solve the bug if we were to add `_setCurrentLocker` in initialize. This is because we would have set the current hook OUTSIDE of a lock so we would never actually clear the current hook because that path would never be hit.

I also wonder if any of this warrants `initialize` needing a lock?

Reply:
Really nice find sara! Definitely agree that (a) is the way to go.

And I think I agree that we should add a lock to `initialize`, and do the exact same currentHook flow for that one too.

If we think that hooks might want to do things in before/afterInitialize (which seems like a reasonable assumption)... it feels right that those actions might add deltas to pools and we should add locking too.